In [1]:
from utils import *
import pandas as pd

pd.set_option('display.max_columns', 50)

## Reading dataset

In [2]:
df_train = pd.read_csv('data/cleaned/train.csv', 
                       parse_dates=['date_of_creation'])

display(df_train.head())

industry  deal_value  weighted_amount date_of_creation  \
0            Restaurants    320506.0        2067263.7       2020-03-29   
1  Construction Services     39488.0         240876.8       2019-07-10   
2      Hospitals/Clinics    359392.0        2407926.4       2019-07-27   
3            Real Estate     76774.0         468321.4       2021-01-30   
4     Financial Services    483896.0              NaN       2019-05-22   

       pitch             lead_revenue fund_category geography  \
0  Product_2         50 - 100 Million    Category 2       USA   
1  Product_2  500 Million - 1 Billion    Category 4     India   
2  Product_1  500 Million - 1 Billion    Category 4       USA   
3  Product_2  500 Million - 1 Billion    Category 3       USA   
4  Product_2         50 - 100 Million    Category 3     India   

                      location         poc_name               designation  \
0           Killeen-Temple, TX  Charlene Werner  Executive Vice President   
1                       Ratlam            rakhi    Chairman/CEO/President   
2  Albany-Schenectady-Troy, NY   Ariel Hamilton      SVP/General Counsel    
3   Mount Vernon-Anacortes, WA      Erin Wilson   CEO/Co-Founder/Chairman   
4                      Shimoga           kavita  Executive Vice President   

            hiring_candidate_role      lead_source level_of_meeting  \
0            Community pharmacist          Website          Level 3   
1          Recruitment consultant           Others          Level 1   
2          Health service manager  Marketing Event          Level 1   
3  Therapist, speech and language    Contact Email          Level 2   
4                   Media planner          Website          Level 2   

                  last_lead_update               internal_poc  \
0                         No track           Davis,Sharrice A   
1  Did not hear back after Level 1             Brown,Maxine A   
2                                ?  Georgakopoulos,Vasilios T   
3  Did not hear back after Level 1             Brown,Maxine A   
4                       Up-to-date              Thomas,Lori E   

                       resource  internal_rating  success_probability  
0                           NaN                3                 73.6  
1                            No                5                 58.9  
2                            No                4                 68.8  
3  We have all the requirements                1                 64.5  
4                            No                4                 62.4

### Sorting data by `date_of_creation`

In [3]:
df_train = df_train.sort_values(by='date_of_creation')
df_train.reset_index(drop=True, inplace=True)

#### Inspecting head and tail of the sorted data

In [4]:
df_train.head()

industry  deal_value  weighted_amount date_of_creation  \
0  Technology Consulting    326636.0        1959816.0       2019-01-01   
1                  Other     64710.0         423850.5       2019-01-01   
2      Hospitals/Clinics    170420.0        1073646.0       2019-01-01   
3                  Banks    480392.0        2906371.6       2019-01-01   
4                  Banks    261734.0        1583490.7       2019-01-01   

       pitch             lead_revenue fund_category geography      location  \
0  Product_1        100 - 500 Million    Category 3     India     Brahmapur   
1  Product_1         50 - 100 Million    Category 4     India     Kavaratti   
2  Product_1  500 Million - 1 Billion    Category 1     India   Muzaffarpur   
3  Product_1  500 Million - 1 Billion    Category 3     India     Kavaratti   
4  Product_1         50 - 100 Million    Category 3     India  Mahabubnagar   

        poc_name                                        designation  \
0     gaytri bai                  Chairman/Chief Innovation Officer   
1         kamini                               SVP/General Counsel    
2           puja  Vice President / GM (04-present) : VP Sales an...   
3         poonam                            CEO/Co-Founder/Chairman   
4  alisha loomba                  Chairman/Chief Innovation Officer   

          hiring_candidate_role      lead_source level_of_meeting  \
0         Editor, commissioning           Others          Level 2   
1  Community development worker  Marketing Event          Level 2   
2                   Hydrologist           Others          Level 2   
3                Office manager           Others          Level 3   
4        Manufacturing engineer  Marketing Event          Level 1   

                  last_lead_update       internal_poc  \
0                more than a month    Dyson,William A   
1                          Pending   Massiah,Gerard F   
2                          Pending  Sutton,Michelle R   
3            More than a week back  Van Arter,Derrick   
4  Did not hear back after Level 1      Charles,Caleb   

                       resource  internal_rating  success_probability  
0                   Deliverable                5                62.70  
1  We have all the requirements                5                61.70  
2                            No                3               107.34  
3                            No                1                61.50  
4                            No                5                26.35

In [5]:
df_train.tail()

industry  deal_value  weighted_amount date_of_creation  \
7002  Materials/Manufacturing     45251.0        257930.70       2021-02-15   
7003       Financial Services     52662.0        284374.80       2021-02-15   
7004              Restaurants    326727.0       2025707.40       2021-02-15   
7005            Conglomerates    487253.0       3142781.85       2021-02-15   
7006    Recreational Products    167888.0        990539.20       2021-02-15   

          pitch             lead_revenue fund_category geography  \
7002  Product_1         50 - 100 Million    Category 2     India   
7003  Product_1        100 - 500 Million    Category 2       USA   
7004  Product_2         50 - 100 Million    Category 3     India   
7005  Product_1  500 Million - 1 Billion    Category 3     India   
7006  Product_1  500 Million - 1 Billion    Category 3       NaN   

                             location       poc_name  \
7002                          Cuttack  khushboo khan   
7003  Greenville-Anderson-Mauldin, SC     Chad Brown   
7004                           Tezpur         ambika   
7005                   Vishakhapatnam        deeksha   
7006                            Ajmer     rakhi soni   

                            designation                 hiring_candidate_role  \
7002            CEO/Co-Founder/Chairman             Medical technical officer   
7003               SVP/General Counsel   English as a second language teacher   
7004                                CEO              Radiographer, diagnostic   
7005  Chairman/Chief Innovation Officer       Commercial/residential surveyor   
7006                                CEO                          Bonds trader   

          lead_source level_of_meeting                 last_lead_update  \
7002           Others          Level 3                          Pending   
7003           Others          Level 1  Did not hear back after Level 1   
7004    Contact Email          Level 3            More than a week back   
7005  Marketing Event          Level 3            More than a week back   
7006           Others          Level 2                More than 2 weeks   

               internal_poc        resource  internal_rating  \
7002  Gaskins Jr,Franklin D             Yes                4   
7003            Green,Candy             Yes                1   
7004     Heidelberg,Andre D     Deliverable                4   
7005            Ryker,David      Not enough                1   
7006     Heidelberg,Andre D  Cannot deliver                4   

      success_probability  
7002                 61.3  
7003                 62.2  
7004                 61.5  
7005                 63.5  
7006                 62.0

## Unique value stats for summarizing the data

In [6]:
df_train_nunique = unique_value_stats(df_train)

display(df_train_nunique)

column  unique_value_count
0                industry                 171
1              deal_value                6907
2         weighted_amount                6480
3        date_of_creation                 777
4                   pitch                   2
5            lead_revenue                   3
6           fund_category                   4
7               geography                   2
8                location                 597
9                poc_name                5261
10            designation                  10
11  hiring_candidate_role                 639
12            lead_source                   4
13       level_of_meeting                   3
14       last_lead_update                  11
15           internal_poc                  60
16               resource                   6
17        internal_rating                   5
18    success_probability                 248

### Summarizing training dataset:

* The data is collected from 01 Jan 2019 to 15 Feb 2021


* There are 171 industries in the data


* Lead was generated for two products i.e. `Product_1` and `Product_2`


* The revenue of the lead's organization can lie in the following range:

```python
['50 - 100 Million', '100 - 500 Million', '500 Million - 1 Billion']
```
    
* Funding category can be `['Category 1', 'Category 2', 'Category 3', 'Category 4']`


* Leads belong to two countries i.e. `USA` and `India`


* Designation has the following values:

```python
['Chairman/Chief Innovation Officer', 'SVP/General Counsel ',
 'Vice President / GM (04-present) : VP Sales and Marketing (01-04)',
 'CEO/Co-Founder/Chairman', 'CEO/Chairman/President',
 'Chairman/CEO/President', 'CEO', 'CEO/President',
 'Chief Executive Officer', 'Executive Vice President'
]
```

Looks like `designation` column has some ambiguities such as `CEO` or `Chief Executive Officer`


* The lead can come from the following sources:

```python
['Contact Email', 'Website', 'Marketing Event', 'Others']
```

* The level of meeting is defined [here](https://github.com/sank3t/Reduce-Marketing-Spend#call-level)


* The lead update has following values:

```python
['more than a month',
 'Pending',
 'More than a week back',
 'Did not hear back after Level 1',
 '?',
 'Up-to-date',
 '2 days back',
 '5 days back',
 'No track',
 'Following up but lead not responding',
 'More than 2 weeks'
 ]
 ```
 
 * The availability of resource can have following values:
 
```python
 ['Deliverable', 'We have all the requirements', 
  'No', 'Not enough',
  'Cannot deliver', 'Yes'
 ]
```

* The lead is rated on a scale of **1 to 5**

#### Interesting fact 🙂

The leads from India have their `poc_name` in lower case.

## Data Preparation

### Fixing the ambiguous values in `designation` column

In [7]:
replace_designation_dict = {
    'Chairman/Chief Innovation Officer': 'CEO',
    'CEO/Co-Founder/Chairman': 'CEO',
    'CEO/Chairman/President': 'CEO',
    'Chairman/CEO/President': 'CEO',
    'CEO/President': 'CEO',
    'Chief Executive Officer': 'CEO',
    'Vice President / GM (04-present) : VP Sales and Marketing (01-04)': 'SVP/General Counsel',
    'Executive Vice President': 'SVP/General Counsel'
}

df_train = df_train.replace({'designation': replace_designation_dict})
df_train['designation'] = df_train['designation'].str.strip()

### Fixing the ambiguous values in `resource` column

In [8]:
replace_resource_dict = {
    'Deliverable': 'Yes',
    'We have all the requirements': 'Yes',
    'Not enough': 'No',
    'Cannot deliver': 'No'
}

df_train = df_train.replace({'resource': replace_resource_dict})
df_train['resource'] = df_train['resource'].str.strip()

In [9]:
df_train.to_csv('data/transformed/train.csv', index=False)